modules/nw-operator-vsphere-multiple-nic.adoc

```diff
@@ -0,0 +1,65 @@
+// Module included in the following assemblies:
+//
+// * installing/installing_vsphere/ipi/installing-vsphere-installer-provisioned-network-customizations.adoc
+// * installing/installing_vsphere/upi/installing-vsphere-network-customizations.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="nw-operator-vsphere-multiple-nic_{context}"]
+= Specifying multiple network interface controllers (NICs)
+
+Before you install an {product-title} cluster on a {vmw-short} host, you can specify multiple network interface controllers (NICs) for a networking implementation so that the {vmw-short} cloud controller manager (CCM) can select the appropriate NIC for a given networking situation. The {vmw-full} can use the NIC for managing pods and services on your cluster. For this configuration, you must specify internal and external Classless Inter-Domain Routing (CIDR) implementations in the {vmw-short} cloud controller manager (CCM) configuration. Each CIDR implementation lists an IP address range that the CCM uses to determine what NICs interact with traffic from internal and external networks.
+
+If you do not configure internal and external CIDR implementations in the {vmw-short} CCM configuration, the {vmw-short} CCM selects the first NIC that the host operating system reports as the primary network interface. This configuration can cause new nodes that associate with a `MachineSet` object with a single NIC to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. In certain situations, the reported network interface might not be the expected network interface. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.
+
+[IMPORTANT]
+====
+To prevent any NIC selection issues, only configure the amount of NICs that you will use. If you configured multiple NICs for a network implementation but you intend to use only a single NIC, ensure that you deploy your cluster nodes on the single NIC.
+
+For a single NIC configuration, you must specify the same IP address for the internal and external CIDR implementations. This configuration ensures that the CCM selects your preferred NIC and not the first NIC that the host operating system reports.
```

Comment:
Thanks. Removed:

> For a single NIC configuration, you must specify the same IP address for the internal and external CIDR implementations. This configuration ensures that the CCM selects your preferred NIC and not the first NIC that the host operating system reports.

Reply:
Added a note before the .yml config:

`For most networking situations, consider setting the standard multiple-NIC configuration. This configuration requires that you set the same IP address ranges in the `nodeNetworking.internal.networkSubnetCidr` and `nodeNetworking.external.networkSubnetCidr` parameters.`

Reply:
I think we have to be careful because even single-NIC has IP selection issues with dual-stack, and we should no longer assume IPv4 single-stack as the default.

We hardcoded the IP address selection workarounds into CCM, but there might still be bugs.  We exclude the `169.254.169.2` and `fd69::2` OVN masquerade networks from Internal and External IP selection.

Also, it's not NIC selection really, is it?   It's Internal and External IP selection per protocol, that's different from NICs.


Maybe:
> To prevent any IP address selection issues for each IP protocol, only configure the amount of IP addresses and IP protocols that you will use. If you configured multiple NICs for a network implementation but you intend to use only the IP protocols on only a single NIC, ensure that you deploy your cluster nodes on the single NIC.



Dualstack logs with many possible IPs:  
```
nodelink_controller.go:109] Adding internal IP "10.93.99.170" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "169.254.169.2" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "fd69::2" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "fd65:a1a8:60ad:1260::5" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "fd65:a1a8:60ad:1260::54" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "fd65:a1a8:60ad:1260:250:56ff:fe85:b9a2" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
nodelink_controller.go:109] Adding internal IP "fe80::250:56ff:fe85:b9a2" for machine "ci-ln-g9pj6h2-c1627-v8n7m-master-1" to indexer
```

Reply:
It might actually be better to always specify `networkSubnetCidr` based on MachineNetwork in all cases.   The hardcode `ExcludeNetworkSubnetCIDR` of `fd69::2/128` is no longer always correct since [SDN-3516](https://issues.redhat.com/browse/SDN-3516) and https://github.com/openshift/openshift-docs/pull/61109 maybe.

Reply:
Thanks, @rbbratta . I updated that note as per your suggestion.

`Re: It might actually be better to always specify networkSubnetCidr based on MachineNetwork in all cases. The hardcode ExcludeNetworkSubnetCIDR of fd69::2/128 is no longer always correct since [SDN-3516](https://issues.redhat.com/browse/SDN-3516) and https://github.com/openshift/openshift-docs/pull/61109 maybe.
`

Should I link to the https://github.com/openshift/openshift-docs/pull/61109/files published docs?

Should I remove the nodeNetworking segment from my ``cluster-infrastructure-02-config.yml` example and replace it with a `MachineNetwork` segement?

Reply:
Okay, so the issue here is actually multiple subnets, regardless of how many NICs are configured. We should probably replace any reference to NICs with subnets to avoid confusion.